### PR TITLE
Feature: Laminas MVC and Mezzio integration

### LINKED ISSUE #1: Provide seamless integration with Laminas MVC and Mezzio apps
We'd like to provide seamless integration with Laminas MVC and Mezzio application,
so that developer can run just:

```bash
$ composer require laminas/laminas-cli
```

and start using the tool:

```bash
$ vendor/bin/laminas-cli
```

### PR DESCRIPTION
**Pull request by michalbundyra**
Resolves #1

### CODE REVIEW

File: docs/book/intro.md [hunk 1]
```diff
@@ -5,3 +5,132 @@ laminas-cli is a command-line interface for Laminas projects.
 It provides seamless integration with Laminas MVC and Mezzio applications.
 
 It is easily extensible by custom command. 
+
+## Integration with component
+
+If you'd like to add integration with laminas-cli into your components
+you have to the following:
+
+1. Use `laminas/laminas-cli` as dev dependency:
```
Comment (weierophinney) [inline docs/book/intro.md]: Why a `dev` dependency? I often have CLI scripts I want to use in production as well, and it seems this is allowing the ability to add arbitrary commands to the tooling?

File: docs/book/intro.md [hunk 2]
```diff
@@ -5,3 +5,132 @@ laminas-cli is a command-line interface for Laminas projects.
...
+
+```php
+namespace MyNamespace\Command;
+
+use Symfony\Component\Console\Command\Command;
+
+class MyCommand extends Command
+{
+    /** @var string */
+    protected static $defaultName = 'my-namespace:my-command';
```
Comment (weierophinney) [inline docs/book/intro.md]: Why this, and not having a map of command name -> service pairs, such as the ContainerCommandLoader uses in symfony/console?

File: docs/book/intro.md [hunk 3]
```diff
@@ -5,3 +5,132 @@ laminas-cli is a command-line interface for Laminas projects.
 It provides seamless integration with Laminas MVC and Mezzio applications.
 
 It is easily extensible by custom command. 
+
+## Integration with component
```
Comment (froschdesign) [inline docs/book/intro.md]: I will wait with a review for the documentation until the implementation is completed.

Comment (michalbundyra) [reply docs/book/intro.md]: Sure, no problem, thanks, @froschdesign :+1:

File: docs/book/intro.md (hunk 2 above)
Comment (michalbundyra) [reply docs/book/intro.md]: Changed in the other PR per suggestion. No longer needed.

File: docs/book/intro.md (hunk 1 above)
Comment (michalbundyra) [reply docs/book/intro.md]: This is instruction for integration with other libraries, not with the application.
In the application it should be installed as non-dev dependency if user needs to run any commands on production.

For other libraries - let say mezzio we'd like to install it as dev dependency and add the commands there. It should be suggested package tho, so if user want to use cli commands should install laminas-cli in the project.

We will add this dependency to require in both skeletons: MVC and Mezzio.

File: src/ContainerResolver.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            $serviceManager->setService('ApplicationConfig', $appConfig);
+
+            $serviceManager->get('ModuleManager')->loadModules();
+
+            return $serviceManager;
+        }
+
+        throw new RuntimeException('Cannot detect DI container');
+    }
+}
```
Comment (weierophinney) [inline src/ContainerResolver.php]:
```suggestion
final class ContainerResolver
{
    /**
     * Try to find container in Laminas application.
     * Supports out of the box Laminas MVC and Mezzio applications.
     *
     * @throws RuntimeException When cannot locate PSR-11 container for the application.
     */
    public static function resolve() : ContainerInterface
    {
        if (file_exists('config/container.php')) {
            return self::resolveDefaultContainer();
        }
        
        if (file_exists('config/application.config.php')
            && class_exists(ServiceManager::class)
        ) {
            return self::resolveMvcContainer();
        }

        throw new RuntimeException('Cannot detect DI container');
    }

    private static function resolveDefaultContainer(): ContainerInterface
    {
        $container = include 'config/container.php';

        if (! $container instanceof ContainerInterface) {
            throw new RuntimeException('Cannot detect DI container');
        }

        return $container;
    }

    private static function resolveMvcContainer(): ContainerInterface
    {
        $appConfig = include 'config/application.config.php';
        if (file_exists('config/development.config.php')) {
            $appConfig = ArrayUtils::merge(
                $appConfig,
                include 'config/development.config.php'
            );
        }

        $smConfig = new ServiceManagerConfig($appConfig['service_manager'] ?? []);

        $serviceManager = new ServiceManager();
        $smConfig->configureServiceManager($serviceManager);
        $serviceManager->setService('ApplicationConfig', $appConfig);

        $serviceManager->get('ModuleManager')->loadModules();

        return $serviceManager;
    }
}
```

I'm suggesting splitting the functionality into multiple methods, to make it more clear what each strategy is and why it gets invoked.

The exception thrown from `resolveDefaultContainer()` should likely change to indicate that it failed to load a PSR-11 container from `config/container.php` if you accept this suggestion.

Comment (michalbundyra) [reply src/ContainerResolver.php]: Changed, thanks for the suggestion.

File: src/ContainerCommandLoader.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @internal
+ */
+final class ContainerCommandLoader extends SymfonyContainerCommandLoader
+{
+    public function get(string $name) : Command
+    {
+        $command = parent::get($name);
+        $command->setName($name);
```
Comment (weierophinney) [inline src/ContainerCommandLoader.php]: I like this A LOT.

It means you can specify a default command name, but the _application_ gets to override it via the configuration map.

Nice solution!

Comment (michalbundyra) [reply src/ContainerCommandLoader.php]: Yes... unless you change the name in `configure`. And this is another problem, maybe not a big problem :)

**Comment by michalbundyra** [comment]
Just noted that tests are failing because `CommandLoaderInterface` adds type hint on params in v5. I think we still want to support v4, so I need to add compatibility layer in that case.

**Pull request merged**